mlir/lib/Dialect/Func/Transforms/AnnotateFunctionType.cpp

```diff
@@ -0,0 +1,48 @@
+//===- AnnotateInputTypes.cpp - Type attribute annotation for func ops ----===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file implements a pass that creates type attributes for func parameters,
+// that mirror the actual type. This is useful when the func op input types
+// might change.
+//
+//===----------------------------------------------------------------------===//
+
+#include "mlir/Dialect/Func/Transforms/Passes.h"
+
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/IR/BuiltinAttributes.h"
+#include "mlir/Pass/Pass.h"
+
+using namespace mlir;
+
+namespace mlir::func {
+#define GEN_PASS_DECL_ANNOTATEFUNCTIONTYPE
```

Comment:
This should go into `mlir/Dialect/Func/Transforms/Passes.h` for other passes to use; only the definition needs to be here.